File: script/worldgen.js
/**
 * World generation
 * @param {World} world - World state object to modify
 */
class Worldgen {
  constructor(world) {
    this.world = world;
  }

  /**
   * Generate a benchmark world with random tiles from the tileset
   * @param {string[]} mask - Array of tile types to use
   */
  generateBenchmarkWorld(mask = Object.keys(TILESET)) {
    this.world.age = 0;
    this.world.surfaceY = this.world.rows - 1;

    // World containing random tiles
    this.world.tiles = [];
    for (let y = 0; y < this.world.rows; y++) {
      let row = [];
      for (let x = 0; x < this.world.cols; x++) {
        // Random tile from mask
        let tile = mask[randomIntInclusive(0, mask.length - 1)];
        row.push(tile);
      }
      this.world.tiles.push(row);
    }
  }

  /**
   * Generate a world
   * All options scale with world size, numbers are for a 10000 tile (100x100) world
   * @param {object} options - World generation options
   * @param {number} options.skyProp - Proportion of world that is sky
   * @param {number} options.startingAge - Starting age of world
   * @param {number} options.startAreaSize - Size of starting area
   * @param {number} options.sandCount - Number of sand patches
   * @param {number} options.sandMinSize - Minimum size of sand patches
   * @param {number} options.sandMaxSize - Maximum size of sand patches
   * @param {number} options.stoneCount - Number of stone patches
   * @param {number} options.stoneMinSize - Minimum size of stone patches
   * @param {number} options.stoneMaxSize - Maximum size of stone patches
   * @param {number} options.waterCount - Number of water patches
   * @param {number} options.waterMinSize - Minimum size of water patches
   * @param {number} options.waterMaxSize - Maximum size of water patches
   * @param {number} options.hollowCount - Number of hollow patches
   * @param {number} options.hollowMinSize - Minimum size of hollow patches
   * @param {number} options.hollowMaxSize - Maximum size of hollow patches
   * @param {number} options.surfacePlantCount - Number of plants on surface
   * @param {number} options.plantCount - Number of plants
   * @param {number} options.fungusCount - Number of fungus patches
   * @param {number} options.fungusMinSize - Minimum size of fungus patches
   * @param {number} options.fungusMaxSize - Maximum size of fungus patches
   * @param {number} options.noiseCount - Number of noise patches
   * @param {number} options.noiseMinSize - Minimum size of noise patches
   * @param {number} options.noiseMaxSize - Maximum size of noise patches
   *
   */
  generate({
    skyProp = 0.2,
    startingAge = 100,
    startAreaSize = 10,
    sandCount = 50,
    sandMinSize = 4,
    sandMaxSize = 10,
    stoneCount = 15,
    stoneMinSize = 4,
    stoneMaxSize = 8,
    waterCount = 8,
    waterMinSize = 5,
    waterMaxSize = 15,
    hollowCount = 15,
    hollowMinSize = 4,
    hollowMaxSize = 10,
    surfacePlantCount = 10,
    plantCount = 300,
    fungusCount = 30,
    fungusMinSize = 1,
    fungusMaxSize = 4,
    noiseCount = 200,
    noiseMinSize = 4,
    noiseMaxSize = 6,
  }) {
    this.world.age = 0;
    this.world.setAge(0);

    const surfaceY = Math.round(this.world.rows * (1 - skyProp));
    this.world.surfaceY = surfaceY;
    const midX = Math.round(this.world.cols / 2);

    // Build 2d tile array
    this.world.tiles = [];
    for (let y = 0; y < this.world.rows; y++) {
      let row = [];
      for (let x = 0; x < this.world.cols; x++) {
        // Default to SOIL underground and AIR above
        let tile = y < surfaceY ? "SOIL" : "AIR";
        row.push(tile);
      }
      this.world.tiles.push(row);
    }
    this.world.setTiles(JSON.stringify(this.world.tiles));

    // Sand
    this._generatePatches(
      sandCount,
      surfaceY,
      sandMinSize,
      sandMaxSize,
      "SAND",
    );

    // Stones
    this._generatePatches(
      stoneCount,
      surfaceY,
      stoneMinSize,
      stoneMaxSize,
      "STONE",
    );

    // Water
    this._generatePatches(
      waterCount,
      surfaceY - waterMaxSize * 2,
      waterMinSize,
      waterMaxSize,
      "WATER",
      ["SOIL", "SAND", "STONE"],
    );

    // Air pockets
    this._generatePatches(
      hollowCount,
      surfaceY,
      hollowMinSize,
      hollowMaxSize,
      "AIR",
      ["SOIL", "SAND", "STONE", "WATER"],
    );

    // Fungus
    this._generatePatches(
      fungusCount,
      surfaceY - fungusMaxSize * 2,
      fungusMinSize,
      fungusMaxSize,
      "FUNGUS",
      ["SOIL", "SAND"],
    );

    // Noise (to make shapes less obvious)
    // These are SOIL patches that can occlude other patches to break up shapes
    this._generatePatches(
      noiseCount,
      surfaceY,
      noiseMinSize,
      noiseMaxSize,
      "SOIL",
      ["SAND", "STONE", "WATER", "FUNGUS", "AIR", "PLANT"],
    );

    // Underground Plant
    this._generatePatches(plantCount, surfaceY, 1, 1, "PLANT", [
      "WATER",
      "AIR",
    ]);

    // Surface Plant
    const plantProb = surfacePlantCount / 100;
    for (let x = 0; x < this.world.cols; x++) {
      if (Math.random() <= plantProb) {
        this.world.setTile(x, this._findSurfaceY(x) + 1, "PLANT");
      }
    }

    // Bedrock
    for (let x = 0; x < this.world.cols; x++) {
      this.world.fillCircle(x, 0, randomIntInclusive(1, 6), "STONE");
    }

    // Starting area
    // Clear a cone shape
    const queenToCeil = this.world.rows - surfaceY + 1;
    const halfStartArea = Math.round(startAreaSize / 2);
    for (let x = midX - halfStartArea; x < midX + halfStartArea; x++) {
      this.world.fillCircle(x, this.world.rows, queenToCeil, "AIR");
    }
    this.world.fillCircle(midX, surfaceY, startAreaSize, "SOIL", [
      "SAND",
      "STONE",
    ]);
    // Guarantee an easy to reach fungus
    this.world.fillCircle(
      randomIntInclusive(midX - halfStartArea, midX + halfStartArea),
      randomIntInclusive(surfaceY - startAreaSize, surfaceY),
      4,
      "FUNGUS",
    );

    for (let i = 0; i < startingAge; i++) {
      this.world.tick();
    }

    // Starting units
    this.world.setTile(midX, surfaceY, "QUEEN");
  }

  /**
   * Generate a number of patches of a given tile type, with a given size range
   * count is scaled based on map size (counts are for a 100x100 10000 tile map)
   * @param {number} count - Number of patches to generate (will be scaled based on map size)
   * @param {number} maxHeight - Maximum height to generate patches at
   * @param {number} minSize - Minimum patch size
   * @param {number} maxSize - Maximum patch size
   * @param {string} tile - Tile type to generate
   * @param {string[]} mask - Tile types allowed to be replaced
   */
  _generatePatches(count, maxHeight, minSize, maxSize, tile, mask) {
    // Scale based on map size (counts are for default 100x100 map)
    const tileCount = this.world.rows * this.world.cols;
    count = (count * tileCount) / 10000;

    for (let i = 0; i < count; i++) {
      this.world.fillCircle(
        randomIntInclusive(0, this.world.cols),
        randomIntInclusive(0, maxHeight),
        randomIntInclusive(minSize, maxSize),
        tile,
        mask,
      );
    }
  }

  /**
   * Returns the surface Y coordinate at the given X coordinate
   * @param {number} x - X coordinate to check
   * @returns {number} - Y coordinate of the first tile below the given x coordinate that is not AIR or WATER
   */
  _findSurfaceY(x) {
    for (let y = this.world.rows - 1; y >= 0; y--) {
      if (!this.world.checkTile(x, y, JSON.stringify(["AIR", "WATER"]))) return y;
    }
    return 0;
  }
}

if (typeof module === 'object') {
  module.exports = { Worldgen };
}

File: src/worldgen.rs
use crate::world::World;
use crate::utils::Utils;
use crate::definitions;
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Clone)]
pub struct Worldgen {
    world: Rc<RefCell<World>>,
}

impl Worldgen {
    pub fn new(world: Rc<RefCell<World>>) -> Worldgen {
        Worldgen {
            world
        }
    }

    pub fn generate_benchmark_world(&self, mask: &[definitions::TileSet]) {
        self.world.borrow_mut().tiles.clear();
        for _ in 0..self.world.borrow().rows {
            let mut row = Vec::new();
            for _ in 0..self.world.borrow().cols {
                let tile_index = Utils::random_int_inclusive(&0, &(mask.len() as i32 - 1)) as usize;
                let tile = mask[tile_index].clone();
                row.push(tile);
            }
            self.world.borrow_mut().tiles.push(row);
        }
    }

    fn generate(
        &mut self, 
        sky_prop: i32,
        starting_age: i32,
        start_area_size: i32,
        sand_count: i32,
        sand_min_size: i32,
        sand_max_size: i32, 
        stone_count: i32, 
        stone_min_size: i32, 
        stone_max_size: i32, 
        water_count: i32, 
        water_min_size: i32, 
        water_max_size: i32, 
        hollow_count: i32, 
        hollow_min_size: i32, 
        hollow_max_size: i32, 
        surface_plant_count: i32, 
        plant_count: i32, 
        fungus_count: i32, 
        fungus_min_size: i32, 
        fungus_max_size: i32, 
        noise_count: i32, 
        noise_min_size: i32, 
        noise_max_size: i32,
    ) {
        self.world.borrow_mut().age = 0;
        let surface_y: i32 = (self.world.borrow().rows as f64 * (1.0 - sky_prop as f64).round()) as i32;
        self.world.borrow_mut().surface_y = Some(surface_y);
        let mid_x = (self.world.borrow().cols as f64 / 2.0).round() as i32;

        // sand
        self.generate_patches(
            &sand_count,
            &surface_y,
            &sand_min_size,
            &sand_max_size,
            &definitions::TileSet::SAND,
            &None,
        );

        // stone
        self.generate_patches(
            &stone_count,
            &surface_y,
            &stone_min_size,
            &stone_max_size,
            &definitions::TileSet::STONE,
            &None,
        );

        // water
        self.generate_patches(
            &water_count,//     let _worldgen = worldgen::Worldgen::new(world.clone());
            &(surface_y - water_max_size * 2),
            &water_min_size,
            &water_max_size,
            &definitions::TileSet::WATER,
            &None,
        );

        // air pockets 
        self.generate_patches(
            &hollow_count,
            &surface_y,
            &hollow_min_size,
            &hollow_max_size,
            &definitions::TileSet::AIR,
            &Some(
                vec![
                    definitions::TileSet::SOIL, 
                    definitions::TileSet::SAND, 
                    definitions::TileSet::STONE, 
                    definitions::TileSet::WATER,
                ],
            ),
        );

        // fungus
        self.generate_patches(
            &fungus_count,
            &(surface_y - water_max_size * 2),
            &fungus_min_size,
            &fungus_max_size,
            &definitions::TileSet::FUNGUS,
            &Some(vec![definitions::TileSet::SOIL, definitions::TileSet::SAND]),
        );

        // Noise (to make shapes less obvious)
        self.generate_patches(
            &noise_count,
            &surface_y,
            &noise_min_size,
            &noise_max_size,
            &definitions::TileSet::SOIL,
            &Some(
                vec![
                    definitions::TileSet::SAND, 
                    definitions::TileSet::STONE, 
                    definitions::TileSet::WATER, 
                    definitions::TileSet::FUNGUS, 
                    definitions::TileSet::AIR, 
                    definitions::TileSet::PLANT,
                ],
            ),
        );

        // underground plant
        self.generate_patches(
            &plant_count,
            &surface_y,
            &1,
            &1,
            &definitions::TileSet::PLANT,
            &Some(
                vec![
                    definitions::TileSet::WATER, 
                    definitions::TileSet::AIR,
                ],
            ),
        );

        // surface plant
        let plant_prob = surface_plant_count as f64 / 100.0;
        for x in 0..self.world.borrow_mut().cols {
            if rand::random::<f64>() <= plant_prob {
                let y = self.find_surface_y(x) + 1;
                self.world.borrow_mut().set_tile(
                    &x, 
                    &y, 
                    &definitions::TileSet::PLANT, 
                    None
                );
            }
        }

        for x in 0..self.world.borrow_mut().cols {
            let radius = Utils::random_int_inclusive(&1, &6);
            self.world.borrow_mut().fill_circle(
                &x, 
                &0, 
                &radius,
                &definitions::TileSet::STONE, 
                &None
            );
        }

        // Starting area
        let queen_to_ceil = self.world.borrow_mut().rows - surface_y as i32 + 1;
        let half_start_area = (start_area_size as f64 / 2.0).round() as i32;

        for x in (mid_x - half_start_area)..(mid_x + half_start_area) {
            self.world.borrow().fill_circle(
                &x, 
                &self.world.borrow().rows, 
                &queen_to_ceil, 
                &definitions::TileSet::AIR, 
                &None,
            );
        }

        // TODO maybe these should just be i32s anyway
        self.world.borrow().fill_circle(
            &mid_x, 
            &surface_y, 
            &start_area_size, 
            &definitions::TileSet::SOIL,
            &Some(vec![definitions::TileSet::SAND, definitions::TileSet::STONE]),
        );

        // Guarantee an easy to reach fungus
        let random_x = Utils::random_int_inclusive(
            &(mid_x - half_start_area),
            &(mid_x + half_start_area),
        );
        let random_y = Utils::random_int_inclusive(
            &(surface_y - start_area_size),
            &surface_y,
        );
        self.world.borrow().fill_circle(
            &random_x, 
            &random_y, 
            &4, 
            &definitions::TileSet::FUNGUS, 
            &None
        );

        // TODO starting age should definitely be an i32
        for _ in 0..starting_age as i32 {
            self.world.borrow_mut().tick();
        }

        // Starting units
        self.world.borrow_mut().set_tile(
            &mid_x, 
            &(surface_y as i32), 
            &definitions::TileSet::QUEEN,
            None,
        );

    }

    pub fn generate_patches(
        &mut self, 
        count: &i32, 
        max_height: &i32, 
        min_size: &i32, 
        max_size: &i32, 
        tile: &definitions::TileSet, 
        mask: &Option<Vec<definitions::TileSet>>,
    ) {
        let tile_count = self.world.borrow().rows * self.world.borrow().cols;
        let count = (count * tile_count) / 10_000;

        for _ in 0..count {
            let x = Utils::random_int_inclusive(&0, &self.world.borrow().cols);
            let y = Utils::random_int_inclusive(&0, &max_height); 
            let size = Utils::random_int_inclusive(&min_size, &max_size);
            self.world.borrow().fill_circle(&x, &y, &size, tile, mask);
        }
    }

    pub fn find_surface_y(&self, x: i32) -> i32 {
        for y in (0..=self.world.borrow().rows).rev() {
            if self.world.borrow().check_tile(&x, &y, Some(&vec![definitions::TileSet::AIR, definitions::TileSet::WATER])) {
                return y;
            }
        }
        0
    }
}
